mobile_verifier/src/heartbeats/mod.rs

```diff
@@ -256,17 +256,54 @@ impl HeartbeatReward {
         self.reward_weight
     }
 
-    pub fn validated<'a>(
+    pub async fn validated<'a>(
         exec: impl sqlx::PgExecutor<'a> + Copy + 'a,
         epoch: &'a Range<DateTime<Utc>>,
         max_distance_to_asserted: u32,
-    ) -> impl Stream<Item = Result<HeartbeatReward, sqlx::Error>> + 'a {
-        sqlx::query_as::<_, HeartbeatRow>(include_str!("valid_heartbeats.sql"))
-            .bind(epoch.start)
-            .bind(epoch.end)
-            .bind(MINIMUM_HEARTBEAT_COUNT)
-            .fetch(exec)
-            .map_ok(move |row| Self::from_heartbeat_row(row, max_distance_to_asserted))
+    ) -> anyhow::Result<impl Stream<Item = HeartbeatReward> + 'a> {
+        let heartbeat_rows =
+            sqlx::query_as::<_, HeartbeatRow>(include_str!("valid_heartbeats.sql"))
+                .bind(epoch.start)
+                .bind(epoch.end)
+                .bind(MINIMUM_HEARTBEAT_COUNT)
+                .fetch(exec)
+                .try_fold(
+                    HashMap::<(PublicKeyBinary, Option<String>), Vec<HeartbeatRow>>::new(),
+                    |mut map, row| async move {
+                        map.entry((row.hotspot_key.clone(), row.cbsd_id.clone()))
+                            .or_default()
+                            .push(row);
+
+                        Ok(map)
+                    },
+                )
+                .await?;
+
+        Ok(
+            futures::stream::iter(heartbeat_rows).map(move |((hotspot_key, cbsd_id), rows)| {
+                let first = rows.first().unwrap();
```

Comment:
there's no chance this could ever be a None?

Reply:
since the map is created from the rows, at least one row has to exist for this to be called on the vec.

Reply:
sure but is there definitely not a world where that query would never return any rows so you wouldn't be passing an empty/default map into this `stream::iter`?

Reply:
The function passed to `map` won't be executed if there are no entries in the iterator.